service/option.go

```diff
@@ -0,0 +1,16 @@
+// Copyright (c) 2022-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package service
+
+type ClientOption func(c *Client) error
+type ClientReconnectCb func(c *Client, attempt int) error
+
+// WithClientReconnectCb lets the caller set an optional callback to be called prior to
+// performing a WebSocket reconnection.
+func WithClientReconnectCb(cb ClientReconnectCb) ClientOption {
+	return func(c *Client) error {
+		c.reconnectCb = cb
+		return nil
+	}
+}
```

Comment:
tbh, this is bending my mind a little -- it's a modifier that takes a cb, and returns a ClientOption, which attaches the cb to the client.  I feel like I just want a method on the client `RegisterReconnectCb`. But maybe you need this way instead?

Reply:
This approach is rather common throughout the codebase (e.g. [here](https://github.com/mattermost/mattermost-server/blob/master/app/options.go)) which is probably the main reason I went for it. It's a clever way to avoid breaking the API and still allow for future extensibility.

We don't strictly need to do this, obviously this could be implemented in a number of other ways but I think it's simple enough once understood and it comes with some bonuses as mentioned before plus doing it this way we ensure the cb is set at initialization time so we don't have to worry about races.